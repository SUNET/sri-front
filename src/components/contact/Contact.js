import React from "react";
import PropTypes from "prop-types";
import { createFragmentContainer } from "react-relay";
import graphql from "babel-plugin-relay/macro";
import { Form, Col } from "react-bootstrap";
import { withTranslation } from "react-i18next";
import NumberFormat from "react-number-format";

import Worklog from "../Worklog";
import Dropdown from "../Dropdown";
import CopyToClipboard from "../CopyToClipboard";
import AppendChild from "../AppendChild";
import ToggleSection, { ToggleHeading, TogglePanel, PanelEditable } from "../../components/ToggleSection";

import "../../style/ModelDetails.scss";

class Contact extends React.PureComponent {
    static propTypes = {
        onChange: PropTypes.func
    };

    render() {
        let { contact, t } = this.props;
        return (
            <>
                <div className="model-section">
                    <Form.Row>
                        <Col>
                            <article>
                                <ToggleSection>
                                    <ToggleHeading>
                                        <h2>{t("contact-details.notes")}</h2>
                                    </ToggleHeading>
                                    <TogglePanel>
                                        <PanelEditable.Consumer>
                                            {(editable) => {
                                                return <span>test {editable.toString()}</span>;
                                            }}
                                        </PanelEditable.Consumer>
                                    </TogglePanel>
                                </ToggleSection>
                            </article>
                            <hr />
                            <article>
                                <ToggleSection>
                                    <ToggleHeading>
                                        <h2>{t("contact-details.general-information")}</h2>
                                    </ToggleHeading>
                                    <TogglePanel>
                                        <PanelEditable.Consumer>
                                            {(editable) => {
                                                return (
                                                    <>
                                                        <div className="table-details">
                                                            <div>
                                                                <div>Title</div>
                                                                <div>Type</div>
                                                                <div>E-mails</div>
                                                                <div>Phone</div>
                                                            </div>
                                                            <div>
                                                                <div>
                                                                    <div>
                                                                        {!editable ? (
                                                                            contact.title
                                                                        ) : (
                                                                            <Form.Group controlId="formGroupEmail">
                                                                                <Form.Control
                                                                                    placeholder="Email"
                                                                                    name="email"
                                                                                    defaultValue={contact.email}
                                                                                    onChange={(e) =>
                                                                                        this.props.onChange(e)
                                                                                    }
                                                                                />
                                                                            </Form.Group>
                                                                        )}
                                                                    </div>
                                                                    <div>
                                                                        {!editable ? (
                                                                            contact.contact_type
                                                                        ) : (
                                                                            <Dropdown
                                                                                emptyLabel="Select type"
                                                                                type="contact_type"
                                                                                onChange={(e) => this.props.onChange(e)}
                                                                                defaultValue={contact.contact_type}
                                                                            />
                                                                        )}
                                                                    </div>
                                                                    <div>
                                                                        {!editable ? (
                                                                            <CopyToClipboard>
                                                                                {contact.email}
                                                                            </CopyToClipboard>
                                                                        ) : (
                                                                            <AppendChild>
                                                                                <Form.Group controlId="formGroupEmail">
                                                                                    <Form.Control
                                                                                        placeholder="Email"
                                                                                        name="email"
                                                                                        defaultValue={contact.email}
                                                                                        onChange={(e) =>
                                                                                            this.props.onChange(e)
                                                                                        }
                                                                                    />
                                                                                </Form.Group>
                                                                            </AppendChild>
                                                                        )}
                                                                    </div>
                                                                    <div>
                                                                        <NumberFormat
                                                                            value={contact.phone}
                                                                            displayType={editable ? "input" : "text"}
                                                                            format="+34 ### ### ###"
                                                                        />
                                                                    </div>
                                                                </div>
                                                            </div>
                                                        </div>
                                                        <div className="table-details">
                                                            <div>
                                                                <div>PGP Fingerprint</div>
                                                            </div>
                                                            <div>
                                                                <div>
                                                                    <Form.Group controlId="formGroupFirstName">
                                                                    <NumberFormat
                                                                        className="auto"
                                                                        value={contact.PGP_fingerprint}
                                                                        displayType={editable ? "input" : "text"}
                                                                        format="#### #### #### #### #### #### #### #### #### ####"
                                                                    />
                                                                </Form.Group>
                                                                </div>
                                                            </div>
                                                        </div>
                                                        <span>test {editable.toString()}</span>
                                                    </>
                                                );
                                            }}
                                        </PanelEditable.Consumer>
                                    </TogglePanel>
                                </ToggleSection>
                            </article>
                            <hr />
                            <article>
                                <ToggleSection>
                                    <ToggleHeading>
                                        <h2>{t("contact-details.profesional-details")}</h2>
                                    </ToggleHeading>
                                    <TogglePanel>
                                        <div className="table-details">
                                            <div>
                                                <div>Role</div>
                                                <div>Organization ID</div>
                                                <div>Organization</div>
                                            </div>
                                            <div>
                                                {contact.roles.map((role, index) => {
                                                    return (
                                                        <div key={index}>
                                                            <div>{role.name}</div>
                                                            <div>{role.end_node.handle_id}</div>
                                                            <div>{role.end_node.name}</div>
                                                        </div>
                                                    );
                                                })}
                                            </div>
                                        </div>
                                    </TogglePanel>
                                </ToggleSection>
                            </article>
                            <Form.Group controlId="formGroupFirstName">
                                <Form.Control
                                    placeholder="First name"
                                    name="first_name"
                                    defaultValue={contact.first_name}
                                    onChange={(e) => this.props.onChange(e)}
                                />
                            </Form.Group>
                            <Form.Group controlId="formGroupLastName">
                                <Form.Control
                                    placeholder="Last name"
                                    name="last_name"
                                    defaultValue={contact.last_name}
                                    onChange={(e) => this.props.onChange(e)}
                                />
                            </Form.Group>
                        </Col>
                    </Form.Row>
                </div>
                <div className="model-section">
                    <article>
                        <Worklog model={contact}/>
                    </article>
                </div>
            </>
        );
    }
}

const ContactFragment = createFragmentContainer(withTranslation()(Contact), {
    contact: graphql`
        fragment Contact_contact on Contact {
            handle_id
            title
            contact_type
            first_name
            last_name
            email
            phone
            PGP_fingerprint
            roles {
                name
                end_node {
                    handle_id
                    name
                }
            }
            comments {
                user {
                    first_name
                    last_name
                }
                comment
                submit_date
            }
        }
    `
});

export default ContactFragment;
