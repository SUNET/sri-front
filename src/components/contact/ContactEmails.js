import { withTranslation } from "react-i18next";
import _EmailPhonesParentClass from "../common/_EmailPhonesParentClass";

class ContactEmails extends _EmailPhonesParentClass {
    ENTITY_NAME = "emails";
    FIELD_NAME = "email";
    WITH_COPY_ROW = true;
    PLACEHOLDER_TRANSLATION = "general-forms/email";
}

export default withTranslation()(ContactEmails);
