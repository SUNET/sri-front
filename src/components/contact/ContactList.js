import React from "react";
import PropTypes from "prop-types";
import { createPaginationContainer } from "react-relay";
import graphql from "babel-plugin-relay/macro";
import { Button } from "react-bootstrap";
import { withRouter } from "react-router-dom";
import { withTranslation } from "react-i18next";

import { ITEMS_PER_PAGE } from "../../constants";
import ContactRow from "./ContactRow";
import FilterColumnsContainer from "../../containers/FilterColumns";

import "../../style/ModelList.scss";

//mock for when the backend is ready
const defaultColumns = [
    { name: "Name", value: "name" },
    { name: "Organization", value: "organization", filter: "order" },
    { name: "Roles", value: "roles" },
    { name: "Contact Type", value: "contact_type" }
];

export class ContactList extends React.PureComponent {
    static propTypes = {
        contacts: PropTypes.object.isRequired
    };

    _loadMore = () => {
        if (!this.props.relay.hasMore()) {
            console.log(`Nothing more to load`);
            return;
        } else if (this.props.relay.isLoading()) {
            console.log(`Request is already pending`);
            return;
        }
        this.props.changeCount(ITEMS_PER_PAGE);
        this.props.relay.loadMore(ITEMS_PER_PAGE);
    };

    handleFilterColumns = () => {};

    _handleOnClick = (event, data) => {
        // Redirection to contact detail
        this.props.history.push(`${this.props.match.url}/${data.handle_id}`);
    };

    renderHeaderList() {
        return (
            <>
                <div></div>
                {defaultColumns.map((column, index) => {
                    // Hiding the columns passed by props
                    if (this.props.columns_visible[column.value] || this.props.all_columns) {
                        return (
                            <div key={index}>
                                {column.name}
                                {column.filter === "order" && (
                                    <FilterColumnsContainer
                                        type="order"
                                        columns={this.props.organization_types.getChoicesForDropdown}
                                    />
                                )}
                            </div>
                        );
                    } else {
                        return null;
                    }
                })}
                <FilterColumnsContainer
                    type="hidden-col"
                    columns={defaultColumns}
                    filterColumns={this.handleFilterColumns}
                />
            </>
        );
    }

    renderList() {
        console.log(this.props);
        let models = this.props.contacts;
        return (
            <>
                {models.contacts.edges.map(({ node }) => {
                    return (
                        <ContactRow
                            key={node.__id}
                            contact={node}
                            onClick={this._handleOnClick}
                            columnsVisible={this.props.columns_visible}
                            showAllColumns={this.props.all_columns}
                        />
                    );
                })}
            </>
        );
    }

    render() {
        const { t } = this.props;
        return (
            <section>
                <div className="model-list">
                    <div>{this.renderHeaderList()}</div>
                    <div>{this.renderList()}</div>
                </div>
                <Button onClick={() => this._loadMore()} variant="outline-primary">
                    {t("paginator.load_more")}
                </Button>
            </section>
        );
    }
}

export default createPaginationContainer(
    withTranslation()(withRouter(ContactList)),
    {
        contacts: graphql`
            fragment ContactList_contacts on Query
                @argumentDefinitions(
                    count: { type: "Int" }
                    cursor: { type: "String" }
                    filter: { type: ContactFilter }
                    orderBy: { type: ContactOrderBy }
                ) {
                contacts(first: $count, after: $cursor, filter: $filter, orderBy: $orderBy)
                    @connection(key: "ContactList_contacts", filters: []) {
                    edges {
                        node {
                            handle_id
                            ...ContactRow_contact
                        }
                    }
                    pageInfo {
                        hasNextPage
                        endCursor
                    }
                }
            }
        `,
        organization_types: graphql`
            fragment ContactList_organization_types on Query {
                getChoicesForDropdown(name: "organization_types") {
                    name
                    value
                }
            }
        `
    },
    {
        direction: "forward",
        query: graphql`
            # Pagination query to be fetched upon calling 'loadMore'.
            # Notice that we re-use our fragment, and the shape of this query matches our fragment spec.
            query ContactListForwardQuery($count: Int!, $cursor: String, $orderBy: ContactOrderBy) {
                ...ContactList_contacts @arguments(count: $count, cursor: $cursor, orderBy: $orderBy)
            }
        `,
        getConnectionFromProps(props) {
            return props.contacts && props.contacts.contacts;
        },
        // This is also the default implementation of `getFragmentVariables` if it isn't provided.
        getFragmentVariables(previousVariables, totalCount) {
            return {
                ...previousVariables,
                count: totalCount
            };
        },
        getVariables(props, paginationInfo, fragmentVariables) {
            return {
                count: paginationInfo.count,
                cursor: paginationInfo.cursor,
                orderBy: fragmentVariables.orderBy
            };
        }
    }
);
