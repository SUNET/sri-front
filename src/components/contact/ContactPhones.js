import { withTranslation } from 'react-i18next';
import _EmailPhonesParentClass from '../common/_EmailPhonesParentClass';

class ContactPhones extends _EmailPhonesParentClass {
  ENTITY_NAME = 'phones';
  FIELD_NAME = 'phone';
  WITH_COPY_ROW = false;
  PLACEHOLDER_TRANSLATION = 'general-forms/phone';
}

export default withTranslation()(ContactPhones);
