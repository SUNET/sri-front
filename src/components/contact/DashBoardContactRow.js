import React from 'react';
import graphql from 'babel-plugin-relay/macro';
import PropTypes from 'prop-types';
import { createFragmentContainer } from 'react-relay';
import { Col, Image } from 'react-bootstrap';
import _DashBoardRowParentClass from '../common/_DashBoardRowParentClass';

export class DashBoardContactRow extends _DashBoardRowParentClass {
  constructor(props) {
    super(props);
    this.MAIN_PROP = 'contact';
    this.TIME_PROP = 'modified';
  }

  renderInfo() {
    const element = this.props[this.MAIN_PROP];
    return (
      <>
        <Col className="dash-board-row__image col-auto">
          <div>
            <Image src={require('../../static/img/profile.png')} roundedCircle img-fluid="true" />
          </div>
        </Col>
        <Col className="px-0 dash-board-row__info">
          <div>
            <div className="dash-board-row__info__name">
              {element.first_name} {element.last_name}
            </div>
            {element.roles && (
              <div className="help-text dash-board-row__info__roles">
                {element.roles.map((role, index) => {
                  return (
                    <span className="dash-board-row__info__roles__row" key={index}>
                      {role.end && (
                        <>
                          {role.name} - {role.end.name}
                        </>
                      )}
                    </span>
                  );
                })}
              </div>
            )}
          </div>
        </Col>
      </>
    );
  }
}

DashBoardContactRow.propTypes = {
  contact: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired,
};

const ContactRowFragment = createFragmentContainer(DashBoardContactRow, {
  contact: graphql`
    fragment DashBoardContactRow_contact on Contact {
      id
      first_name
      last_name
      modified
      roles {
        name
        end {
          name
        }
      }
      member_of_groups {
        name
      }
    }
  `,
});

export default ContactRowFragment;
