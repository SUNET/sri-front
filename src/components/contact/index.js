export { default as Contact } from "./Contact";
export { default as ContactDetails } from "./ContactDetails";
export { default as ContactList } from "./ContactList";
export { default as ContactRow } from "./ContactRow";
export { default as CreateContact } from "./CreateContact";
