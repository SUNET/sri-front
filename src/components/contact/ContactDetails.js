import React from "react";
import PropTypes from "prop-types";
import { QueryRenderer } from "react-relay";
import graphql from "babel-plugin-relay/macro";
import { Row, Col, Form } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faStar } from "@fortawesome/free-solid-svg-icons";
import { withTranslation } from "react-i18next";

import Contact from "./Contact";
import EditFieldNoRedux from "../EditFieldNoRedux";
import DeleteContactMutation from "../../mutations/DeleteContactMutation";
import UpdateContactMutation from "../../mutations/UpdateContactMutation";
import environment from "../../createRelayEnvironment";
import InfoCreatorModifier from "../InfoCreatorModifier";

const ContactDetailsQuery = graphql`
    query ContactDetailsQuery($contactId: Int!) {
        getContactById(handle_id: $contactId) {
            ...Contact_contact
            name
            contact_type
            created
            creator {
                email
            }
            modified
            modifier {
                email
            }
        }
    }
`;

class ContactDetails extends React.Component {
    static propTypes = {
        match: PropTypes.shape({
            params: PropTypes.shape({
                id: PropTypes.node
            }).isRequired
        }).isRequired
    };

    constructor(props) {
        super(props);

        this.state = {
            name: "",
            first_name: "",
            last_name: "",
            email: "",
            phone: "",
            contact_type: ""
        };
    }

    UNSAFE_componentWillUpdate(nextProps, nextState) {
        if (this.state.first_name !== nextState.first_name && this.state.last_name !== nextState.last_name) {
            this._handleUpdate(nextState);
        }
    }

    _handleContactChange = (event) => {
        if (event.target.name === "full-name") {
            let fullName = event.target.value;
            fullName = fullName.split(" ");
            this.setState({ first_name: fullName[0], last_name: fullName[1] });
        }
    };

    _handleUpdate = (contact) => {
        const update_contact = {
            id: this.props.match.params.contactId,
            first_name: contact.first_name,
            last_name: contact.last_name,
            email: contact.email,
            phone: contact.phone,
            contact_type: "person"
        };
        UpdateContactMutation(update_contact, environment);
    };

    _handleDelete = () => {
        const contactId = this.props.match.params.contactId;
        DeleteContactMutation(contactId, () => this.props.history.push(`/community/contacts`));
    };

    render() {
        let { t } = this.props;
        return (
            <QueryRenderer
                environment={environment}
                query={ContactDetailsQuery}
                variables={{
                    contactId: this.props.match.params.contactId
                }}
                render={({ error, props }) => {
                    if (error) {
                        return <div>{error.message}</div>;
                    } else if (props) {
                        this.contact = props.getContactById;
                        return (
                            <section className="model-details">
                                <Row>
                                    <Col>
                                        <div className="title-section">
                                            <button
                                                onClick={() => this.props.history.goBack()}
                                                className="btn btn-back outline"
                                            >
                                                <span>{t("actions.back")}</span>
                                            </button>
                                            <EditFieldNoRedux onChange={this._handleContactChange} reduxForm={false}>
                                                <h1>{props.getContactById.name}</h1>
                                            </EditFieldNoRedux>
                                            <FontAwesomeIcon icon={faStar} />
                                        </div>
                                    </Col>
                                    <Col>
                                        <InfoCreatorModifier model={props.getContactById} />
                                    </Col>
                                </Row>
                                <Row>
                                    <Col>
                                        <Form>
                                            <Contact
                                                onChange={this._handleContactChange}
                                                contact={props.getContactById}
                                            />
                                            <div className="text-right mt-4">
                                                <button
                                                    type="button"
                                                    onClick={() => this._handleDelete()}
                                                    className="btn link"
                                                >
                                                    {t("actions.delete")}
                                                </button>
                                                <button
                                                    onClick={() => this._handleUpdate(props.getContactById)}
                                                    className="btn primary lg"
                                                >
                                                    {t("actions.save")}
                                                </button>
                                            </div>
                                        </Form>
                                    </Col>
                                </Row>
                            </section>
                        );
                    }
                    return <div>Loading</div>;
                }}
            />
        );
    }
}

export default withTranslation()(ContactDetails);
