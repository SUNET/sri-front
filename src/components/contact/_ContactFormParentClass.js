// Common imports
import React from "react";
import { FieldArray, Field } from "redux-form";
import { Form, Col } from "react-bootstrap";
// components
import BackCTA from "../common/BackCTA";
import ContactEmails from "../contact/ContactEmails";
import ContactPhones from "../contact/ContactPhones";
import Dropdown from "../Dropdown";
import EditField from "../EditField";
import FieldArrayOrganizationsContact from "../contact/FieldArrayOrganizationsContact";
import FieldInput from "../FieldInput";
import InfoCreatorModifier from "../InfoCreatorModifier";
import SaveCancelCTAs from "../common/SaveCancelCTAs";
import ToggleSection, { ToggleHeading, TogglePanel } from "../../components/ToggleSection";
import Worklog from "../Worklog";
// const
import { isBrowser, isMobile } from "react-device-detect";
// scss
import "../../style/ModelDetails.scss";

const renderFormBlockSection = (editable, data, uniqueKey) => {
    const isPresentState = !editable;
    const presentContent = data.presentContent || "";
    return (
        <div className="form-internal-block__section" key={uniqueKey}>
            <div className="form-internal-block__section__title">{data.title}</div>
            <div
                className={`form-internal-block__section__content ${
                    editable ? "form-internal-block__section__content--edition-mode" : ""
                }`}
            >
                {isPresentState ? presentContent : data.editContent}
            </div>
        </div>
    );
};

class _ContactFormParentClass extends React.Component {
    // GLOBAL VARs
    IS_UPDATED_FORM = false;
    FORM_ID;
    MODEL_NAME = "contact";

    componentDidMount() {
        if (this.IS_UPDATED_FORM) {
            this.updateBreadcrumbsData();
        }
    }

    componentWillUnmount() {
        if (this.IS_UPDATED_FORM) {
            this.props.getOutOfDetails();
        }
    }

    updateBreadcrumbsData() {
        this.props.moveToDetails({
            id: this.props.initialValues.id,
            name: `${this.props.initialValues.first_name} ${this.props.initialValues.last_name} `,
        });
    }

    // Methods
    refetch = () => {
        throw new Error("This method should be overwritten in the child class");
    };
    handleSubmit = () => {
        throw new Error("This method should be overwritten in the child class");
    };
    onClickDelete = () => {
        this.props.onDelete();
    };
    onClickCancel = () => {
        const { shownInModal, history, hideContactModal } = this.props;
        if (shownInModal) {
            hideContactModal();
        } else {
            history.goBack();
        }
    };

    // Common sections RENDERS
    renderEditButton() {
        const { t } = this.props;
        const desktopClass = isBrowser ? "with-vertical-separator with-vertical-separator--right" : "";
        return (
            <div className={`title-section__right-block__buttons ${desktopClass} mr-3`}>
                <button
                    type="button"
                    onClick={() => {
                        this.setState({ editMode: !this.state.editMode });
                    }}
                    className="btn outline btn-edit"
                >
                    <i className="icon-pencil"></i>
                    <span>{t("actions.edit")}</span>
                </button>
            </div>
        );
    }
    renderSaveCancelButtons() {
        const { t } = this.props;
        const textToButtons = this.IS_UPDATED_FORM ? t("actions.delete") : t("actions.cancel");
        const functionToCancel = this.IS_UPDATED_FORM ? this.onClickDelete : this.onClickCancel;
        return <SaveCancelCTAs formId={this.FORM_ID} cancelText={textToButtons} onCancel={functionToCancel} />;
    }
    renderHeader(editMode = true, showBackButton = true, shownInModal = false) {
        return (
            <Form.Row>
                <Col className={`d-inline-flex align-items-center ${isMobile ? "mb-3" : ""}`}>
                    {this.renderHeaderName(editMode, showBackButton, shownInModal)}
                </Col>
                {this.IS_UPDATED_FORM && <Col>{this.renderHeaderRight(shownInModal)}</Col>}
            </Form.Row>
        );
    }
    renderHeaderName(editMode = true, showBackButton = true, shownInModal = false) {
        const editionModeClass = editMode ? "title-section__name-inputs--edition-mode" : "";
        return (
            <div className="title-section">
                {showBackButton && <BackCTA onClick={() => this.props.history.goBack()} />}
                {this.IS_UPDATED_FORM && isMobile && !shownInModal && this.renderEditButton()}
                <div className="vertical-separator"></div>
                <div className={`title-section__name-inputs ${editionModeClass}`}>
                    {this.renderInputName("first_name", editMode)}
                    {this.renderInputName("last_name", editMode)}
                </div>
            </div>
        );
    }
    renderHeaderRight(shownInModal = false) {
        return (
            <div className="title-section__right-block">
                {isBrowser && !shownInModal && this.renderEditButton()}
                <InfoCreatorModifier model={this.props[this.MODEL_NAME]} />
            </div>
        );
    }
    renderInputName(kindOfName, editMode = true) {
        // INFO: kindOfName = 'first_name' || 'last_name'
        const { t, formSyncErrors, fields, form, dispatch } = this.props;
        const placeHolderString =
            kindOfName === "first_name" ? t("contact-details.name") : t("contact-details.lastName");
        return (
            <EditField
                error={formSyncErrors[kindOfName]}
                meta={fields[kindOfName]}
                form={form}
                dispatch={dispatch}
                editable={editMode}
                placeholder={placeHolderString}
                name={kindOfName}
            >
                <h1>{this.props[kindOfName]}</h1>
            </EditField>
        );
    }
    renderWorkLog() {
        const { t, contact } = this.props;
        return (
            <section className="model-section">
                {this.IS_UPDATED_FORM ? (
                    <Worklog model={contact} refetch={this.refetch} />
                ) : (
                    <ToggleSection defaultEditable={false}>
                        <ToggleHeading>
                            <h2>{t("contact-details.worklog")}</h2>
                        </ToggleHeading>
                        <TogglePanel>
                            <Field
                                name="comment"
                                component={FieldInput}
                                as="textarea"
                                rows="3"
                                placeholder={t("worklog.add-comment")}
                            />
                        </TogglePanel>
                    </ToggleSection>
                )}
            </section>
        );
    }

    // Specific toggle sections RENDERS
    renderNotesToggleSection(editMode = true) {
        const { t } = this.props;
        return (
            <ToggleSection>
                <ToggleHeading>
                    <h2>{t("contact-details.notes")}</h2>
                </ToggleHeading>
                <TogglePanel>
                    {editMode ? (
                        <Field
                            name="notes"
                            component={FieldInput}
                            as="textarea"
                            rows="3"
                            placeholder={t("contact-details.add-notes")}
                        />
                    ) : (
                        <span className="pre-text">{this.props.notes}</span>
                    )}
                </TogglePanel>
            </ToggleSection>
        );
    }
    renderGeneralInfoToggleSection(editMode = true) {
        const { t, title, contactTypeObj, pgp_fingerprint } = this.props;
        const generalInfoFirstRow = [
            {
                title: "Title",
                presentContent: title,
                editContent: (
                    <Form.Group>
                        <Field type="text" component={FieldInput} placeholder="Type title" name="title" />
                    </Form.Group>
                )
            },
            {
                title: "Type",
                presentContent: contactTypeObj ? contactTypeObj.name : undefined,
                editContent: (
                    <Dropdown
                        className="auto"
                        emptyLabel="Select type"
                        name="contact_type"
                        type="contact_type"
                        onChange={(e) => {}}
                    />
                )
            },
            {
                title: "E-mails",
                presentContent: (
                    <FieldArray
                        name="emails"
                        t={t}
                        component={ContactEmails}
                        editable={editMode}
                        dispatch={this.props.dispatch}
                    />
                ),
                editContent: (
                    <FieldArray
                        name="emails"
                        t={t}
                        component={ContactEmails}
                        editable={editMode}
                        dispatch={this.props.dispatch}
                    />
                )
            },
            {
                title: "Phone",
                presentContent: (
                    <FieldArray
                        name="phones"
                        t={t}
                        component={ContactPhones}
                        editable={editMode}
                        dispatch={this.props.dispatch}
                    />
                ),
                editContent: (
                    <FieldArray
                        name="phones"
                        t={t}
                        component={ContactPhones}
                        editable={editMode}
                        dispatch={this.props.dispatch}
                    />
                )
            }
        ];
        return (
            <ToggleSection defaultEditable={false}>
                <ToggleHeading>
                    <h2>{t("contact-details.general-information")}</h2>
                </ToggleHeading>
                <TogglePanel>
                    <>
                        <div className="form-internal-block">
                            {generalInfoFirstRow.map((formData, index) => {
                                return renderFormBlockSection(editMode, formData, index);
                            })}
                        </div>
                        <div className="table-details mt-4">
                            <div>
                                <div>PGP Fingerprint</div>
                            </div>
                            <div>
                                <div>
                                    {!editMode ? (
                                        pgp_fingerprint
                                    ) : (
                                        <Form.Group>
                                            <Field
                                                type="text"
                                                component={FieldInput}
                                                className={`${isBrowser ? "xlg" : "auto"}`}
                                                placeholder={t("contact-details.pgp-fingerprint")}
                                                name="pgp_fingerprint"
                                            />
                                        </Form.Group>
                                    )}
                                </div>
                            </div>
                        </div>
                    </>
                </TogglePanel>
            </ToggleSection>
        );
    }
    renderProfessionalDetails(editMode = true) {
        const { t } = this.props;
        return (
            <ToggleSection>
                <ToggleHeading>
                    <h2>{t("contact-details.professional-details")}</h2>
                </ToggleHeading>
                <TogglePanel>
                    <FieldArray
                        name="organizations"
                        component={FieldArrayOrganizationsContact}
                        editable={editMode}
                        dispatch={this.props.dispatch}
                        errors={this.props.formSyncErrors.organizations}
                        metaFields={this.props.fields}
                    />
                </TogglePanel>
            </ToggleSection>
        );
    }
    renderModelMainSection(editMode = true) {
        return (
            <section className="model-section">
                <Form.Row>
                    <Col>
                        {this.renderNotesToggleSection(editMode)}
                        <hr />
                        {this.renderGeneralInfoToggleSection(editMode)}
                        <hr />
                        {this.renderProfessionalDetails(editMode)}
                    </Col>
                </Form.Row>
            </section>
        );
    }

    // Main RENDER
    render() {
        console.error("This method should be overwritten in the child class");
        return <div>This method should be overwritten in the child class</div>;
    }
}

export default _ContactFormParentClass;
