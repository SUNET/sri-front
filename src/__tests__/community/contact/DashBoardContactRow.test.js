import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { DashBoardContactRow } from '../../../components/contact/DashBoardContactRow';

let DashBoardContactRowWrapper;
let DashBoardContactRowComponent;

const FIRST_NAME = 'Nonnah';
const LAST_NAME = 'Bottini';
const ROLE = 'Senior Editor';
const ORG = 'Tavu';
const MODIFIED_DATE_STRING = '2020-05-05T14:53:50.620447+00:00';
const SINCE_TIME_STRING = '18 hours ago';
const defaultProps = {
  onClick: jest.fn(),
  contact: {
    id: 'Q29udGFjdDo4NTU=',
    first_name: FIRST_NAME,
    last_name: LAST_NAME,
    modified: MODIFIED_DATE_STRING,
    roles: [
      {
        name: ROLE,
        end: {
          name: ORG,
        },
      },
    ],
    member_of_groups: [],
  },
};

jest.mock('moment', () => {
  const mMoment = {
    locale: jest.fn().mockReturnThis(),
    fromNow: () => '18 hours ago',
  };
  return jest.fn(() => mMoment);
});

const CreateDashBoardContactRowWrapper = (props) => {
  const allProps = { ...defaultProps, ...props };
  return (
    <div>
      <DashBoardContactRow {...allProps} />
    </div>
  );
};

const mountDashBoardContactRowWrapper = (props) => mount(<CreateDashBoardContactRowWrapper {...props} />);

beforeEach(() => {
  jest.clearAllMocks();
  DashBoardContactRowWrapper = mountDashBoardContactRowWrapper();
  DashBoardContactRowComponent = DashBoardContactRowWrapper.find('DashBoardContactRow');
});
afterEach(() => {
  DashBoardContactRowWrapper.unmount();
});

describe('DashBoardContactRow Component', () => {
  it('Renders correctly', () => {
    expect(toJson(DashBoardContactRowWrapper)).toMatchSnapshot();
    expect(DashBoardContactRowComponent.find('.dash-board-row')).toHaveLength(1);
  });
});
describe('Elements', () => {
  it('Appears Logo Contact in fist COL', () => {
    const wrapper = DashBoardContactRowComponent.find('.dash-board-row');
    const logoContainer = wrapper.find('.dash-board-row__image.col');
    expect(logoContainer).toHaveLength(1);
    const imgElement = logoContainer.find('img');
    expect(imgElement.prop('src')).toBe('profile.png');
  });
  it('Appears First Name and Last Name', () => {
    const wrapperCol = DashBoardContactRowComponent.find('.dash-board-row__info');
    const nameElement = wrapperCol.find('.dash-board-row__info__name');
    expect(nameElement).toHaveLength(1);
    expect(nameElement.text()).toEqual(`${FIRST_NAME} ${LAST_NAME}`);
  });
  it('Appears Role and Organization', () => {
    const wrapperCol = DashBoardContactRowComponent.find('.dash-board-row__info__roles');
    const rolesElements = wrapperCol.find('.dash-board-row__info__roles__row');
    expect(rolesElements).toHaveLength(1);
    expect(rolesElements.find('span').text()).toEqual(`${ROLE} - ${ORG}`);
  });
  it('Appears time since modification', () => {
    const wrapperCol = DashBoardContactRowComponent.find('div.dash-board-row__modified_time');
    expect(wrapperCol).toHaveLength(1);
    expect(wrapperCol.find('div').text()).toEqual(`${SINCE_TIME_STRING}`);
  });
});

describe('Events', () => {
  it('When we click it launches the right method', () => {
    const wrapper = DashBoardContactRowComponent.find('.dash-board-row');
    wrapper.simulate('click');
    expect(defaultProps.onClick.mock.calls).toHaveLength(1);
  });
});
