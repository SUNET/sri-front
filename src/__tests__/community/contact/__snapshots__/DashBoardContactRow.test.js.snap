// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashBoardContactRow Component Renders correctly 1`] = `
<CreateDashBoardContactRowWrapper>
  <div>
    <DashBoardContactRow
      contact={
        Object {
          "first_name": "Nonnah",
          "id": "Q29udGFjdDo4NTU=",
          "last_name": "Bottini",
          "member_of_groups": Array [],
          "modified": "2020-05-05T14:53:50.620447+00:00",
          "roles": Array [
            Object {
              "end": Object {
                "name": "Tavu",
              },
              "name": "Senior Editor",
            },
          ],
        }
      }
      onClick={[MockFunction]}
    >
      <article
        className="dash-board-row"
        onClick={[Function]}
      >
        <Row
          noGutters={false}
        >
          <div
            className="row"
          >
            <Col
              className="dash-board-row__image col-auto"
            >
              <div
                className="dash-board-row__image col-auto col"
              >
                <div
                  className="dash-board-row__image__default"
                >
                   
                </div>
              </div>
            </Col>
            <Col
              className="px-0 dash-board-row__info"
            >
              <div
                className="px-0 dash-board-row__info col"
              >
                <div>
                  <div
                    className="dash-board-row__info__name"
                  >
                    Nonnah
                     
                    Bottini
                  </div>
                  <div
                    className="help-text dash-board-row__info__roles"
                  >
                    <span
                      className="dash-board-row__info__roles__row"
                      key="0"
                    >
                      Senior Editor
                       - 
                      Tavu
                    </span>
                  </div>
                </div>
              </div>
            </Col>
            <Col
              className="col-md-auto dash-board-row__modified_time"
            >
              <div
                className="col-md-auto dash-board-row__modified_time col"
              >
                <span>
                  18 hours ago
                </span>
              </div>
            </Col>
          </div>
        </Row>
      </article>
    </DashBoardContactRow>
  </div>
</CreateDashBoardContactRowWrapper>
`;
